internal/app/go-filecoin/plumbing/cst/chain_state.go

```diff
@@ -211,10 +217,59 @@ func (chn *ChainStateReadWriter) ChainExport(ctx context.Context, head block.Tip
 // ChainImport imports a chain from `in`.
 func (chn *ChainStateReadWriter) ChainImport(ctx context.Context, in io.Reader) (block.TipSetKey, error) {
 	logStore.Info("starting CAR file import")
-	headKey, err := chain.Import(ctx, newCarStore(chn.cst), in)
+	headKey, err := chain.Import(ctx, newCarStore(chn.bstore), in)
 	if err != nil {
 		return block.UndefTipSet.Key(), err
 	}
 	logStore.Infof("imported CAR file with head: %s", headKey)
 	return headKey, nil
 }
+
+// ChainStateTree returns the state tree as a slice of IPLD nodes at the passed stateroot cid `c`.
+func (chn *ChainStateReadWriter) ChainStateTree(ctx context.Context, c cid.Cid) ([]format.Node, error) {
+	return newChainStateCollector(chn.bstore).collectState(ctx, c)
+}
+
+func newChainStateCollector(bs blockstore.Blockstore) *chainStateCollector {
+	offl := offline.Exchange(bs)
+	blkserv := bserv.New(bs, offl)
+	dserv := dag.NewDAGService(blkserv)
+	return &chainStateCollector{
+		dagserv: dserv,
+	}
+}
+
+type chainStateCollector struct {
+	dagserv format.DAGService // Provides access to state tree.
+	state   []format.Node
+}
+
+func (csc *chainStateCollector) collectState(ctx context.Context, stateRoot cid.Cid) ([]format.Node, error) {
+	dagNd, err := csc.dagserv.Get(ctx, stateRoot)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to load stateroot from dagservice %s", stateRoot)
+	}
+	csc.addState(dagNd)
+	seen := cid.NewSet()
+	for _, l := range dagNd.Links() {
+		if err := dag.Walk(ctx, csc.getLinks, l.Cid, seen.Visit); err != nil {
```

Comment:
Does this have a side effect? It looks like the purpose is to validate. If that's the case, this should be `collectAndValidateState` or just `validateState`. If not, the side effect should be documented in a comment.

Reply:
Its purpose is to recursively walk the state tree starting at `stateroot` and call `getLinks` for each IPLD node it visits, `getLinks` adds each node to the slice of IPLD nodes in the chainStateCollector called `state`. Visiting each node should not have any side effects, I will leave a comment at the top of this method stating its intent.